aips/aip-x.md

```diff
@@ -0,0 +1,151 @@
+
+## Impact
+
+Most of the changes will be in our SDK interfaces. This should significantly improve developer's experience to create transaction that invokes multiple Move function calls all at once. This would be particularly helpful for our defi developers as they would be able to implement features such as withdraw coin and put them into a lending protocol in one transaction.
+
+## Alternative solutions
+
+An alternative approach is to introduce a third payload type in our system and have the node software interpret this payload directly. While this could offer some performance improvements by bypassing the full bytecode verifier, it would require reimplementing several critical checks, such as type safety, ability constraints, and reference safety, that are already handled by the verifier. Duplicating this logic at the node level not only adds redundancy but also increases the risk of introducing security vulnerabilities.
+
+## Specification and Implementation Details
+
+
+We will be introducing the following APIs in the ts-sdk:
+
+```
+export type InputBatchedFunctionData = {
+  function: MoveFunctionId;
+  typeArguments?: Array<TypeArgument>;
+  /* Function argument could either be existing entry function argument or come from the output of a previous function */
+  functionArguments: Array<EntryFunctionArgumentTypes | BatchArgument | SimpleEntryFunctionArgumentTypes>;
+};
+
+export class AptosIntentBuilder {
+    /* Move function calls can return values. Those values can be passed to other functions by move, copy or reference */
+    async add_batched_calls(input: InputBatchedFunctionData): Promise<BatchArgument[]>;
+}
+
+/* Existing TransactionBuilder */
+export class Build {
+    async batched_intents(args: {
+        sender: AccountAddressInput;
+        builder: (builder: AptosIntentBuilder) => Promise<AptosIntentBuilder>;
+        options?: InputGenerateTransactionOptions;
+        withFeePayer?: boolean;
+    }): Promise<SimpleTransaction>
+}
+```
+
+With those new apis, developers can pass in a closure that builds up the Move call chains into one transaction. Here's an example:
+
+```
+const transaction = await _aptos.aptos.transaction.build.batched_intents({
+    sender: singleSignerED25519SenderAccount.accountAddress,
+    builder: async (builder) => {
+    /* return_1 should contain the withdrawed Coin<AptosCoin> */
+    let return_1 = await builder.add_batched_calls({
+        function: `0x1::coin::withdraw`,
+        functionArguments: [BatchArgument.new_signer(0), 1],
+        typeArguments: ["0x1::aptos_coin::AptosCoin"]
+    });
+
+    /* return_2 should contain the converted FungibleAsset */
+    let return_2 = await builder.add_batched_calls({
+        function: `0x1::coin::coin_to_fungible_asset`,
+        /* Pass the withdrawed Coin<AptosCoin> to the coin_to_fungible_asset */
+        functionArguments: [return_1[0]],
```

Comment:
it seems a bit odd to assume tuple is returned, and require [0] always.
would it be better to have return_1 by default, and if it is tuple you can specify return_1.tuple_index(0) or something?